frontend/public/components/storage/attach-storage.tsx

```diff
@@ -40,113 +27,131 @@ const PVCDropdown: React.SFC<PVCDropdownProps> = props => {
+  const [selectedContainers, setSelectedContainers] = React.useState([]);
+  const [volumeAlreadyMounted, setVolumeAlreadyMounted] = React.useState(false);
+  const [error, setError] = React.useState('');
+  const [showCreatePVC, setShowCreatePVC] = React.useState('existing');
+  const [newPVCObj, setNewPVCObj] = React.useState(null);
 
-  componentDidMount() {
-    const { kindObj, name, namespace } = this.props;
-    const supportedKinds = [
-      'Deployment',
-      'DeploymentConfig',
-      'ReplicaSet',
-      'ReplicationController',
-      'StatefulSet',
-      'DaemonSet',
-    ];
-
-    if (!kindObj || !_.includes(supportedKinds, kindObj.kind)) {
-      this.setState({ error: 'Unsupported kind.' });
-      return;
-    }
+  const { kindObj, resourceName, namespace } = props;
+  const supportedKinds = [
+    'Deployment',
+    'DeploymentConfig',
+    'ReplicaSet',
+    'ReplicationController',
+    'StatefulSet',
+    'DaemonSet',
+  ];
 
+  React.useEffect(() => {
     // Get the current resource so we can add to its definition
-    k8sGet(kindObj, name, namespace).then(
-      resourceObj => {
-        this.setState({ resourceObj });
-      },
-      err => this.setState({ error: err.message })
-    );
+    k8sGet(kindObj, resourceName, namespace).then(setObj);
+  }, [kindObj, resourceName, namespace]);
+
+  if (!kindObj || !_.includes(supportedKinds, kindObj.kind)) {
+    setError('Unsupported kind.');
+    return;
   }
 
-  updateVolumeName = (claimName: string) => {
-    const { resourceObj } = this.state;
+  const updateVolumeName = (newClaimName: string) => {
     // Check if there is already a volume for this PVC.
-    const volumes = _.get(resourceObj, 'spec.template.spec.volumes');
+    const volumes = _.get(obj, 'spec.template.spec.volumes');
     const volume = _.find(volumes, {
       persistentVolumeClaim: {
-        claimName,
+        claimName: newClaimName,
       },
     }) as any;
 
-    const volumeName = volume ? volume.name : claimName;
-    const volumeAlreadyMounted = !!volume;
-    this.setState({ volumeName, volumeAlreadyMounted });
+    const newVolumeName = volume ? volume.name : newClaimName;
+    const newVolumeAlreadyMounted = !!volume;
+    setVolumeName(newVolumeName);
+    setVolumeAlreadyMounted(newVolumeAlreadyMounted);
+  };
+
+  const handleShowCreatePVCChange: React.ReactEventHandler<HTMLInputElement> = event => {
+    setShowCreatePVC(event.currentTarget.value);
+  };
+
+  const handleSelectContainers = () => {
+    setUseContainerSelector(!useContainerSelector);
+    setSelectedContainers([]);
+    setAllContainers(true);
   };
 
-  handleChange: React.ReactEventHandler<HTMLInputElement> = event => {
-    const { name, value } = event.currentTarget;
-    this.setState({ [name]: value } as any);
+  const handleContainerSelectionChange = (checked, event) => {
+    checked ? selectedContainers.push(event.currentTarget.id) : _.pull(selectedContainers, event.currentTarget.id);
```

Comment:
@zherman0 I don't believe this will update the state and cause a re-render as you are not mutating the selectedContainers by push/pull. If you create a new array instance the state comparison will show two different states and update state for you. To get around this, one example is:

```let mutatedArray = selectedContainers.slice(0);
    checked ? mutatedArray.push(event.currentTarget.id) : _.pull(mutatedArray, event.currentTarget.id);
    setSelectedContainers(mutatedArray);
    setAllContainers(mutatedArray.length === 0);```
After this sample code, the checkbox state works as expected.

Reply:
Thanks!